REGRESSION (299927@main): [ macOS ]TestWebKitAPI.UnifiedPDF.PasswordFormShouldDismissAfterNavigation is a consistent failure

cdumez opened the pull request:
#### ce989ed6eac830b4b32ad6f734f7893adb53e39e
<pre>
REGRESSION (299927@main): [ macOS ]TestWebKitAPI.UnifiedPDF.PasswordFormShouldDismissAfterNavigation is a consistent failure
<a href="https://bugs.webkit.org/show_bug.cgi?id=298987">https://bugs.webkit.org/show_bug.cgi?id=298987</a>
<a href="https://rdar.apple.com/160728656">rdar://160728656</a>

Reviewed by Per Arne Vollan and Basuke Suzuki.

The code in Navigation::innerDispatchNavigateEvent() was getting called from
FrameLoader::commitProvisionalLoad() and would end up constructing the JS
global object too early and the test&apos;s injected bundle would fail to setup
`window.internals`. The reason innerDispatchNavigateEvent() would end up
constructing the JS global object was due to constructing a dummy resolved
JS promise, for the sole purpose of scheduling work in a microtask.
Instead of constructing a JS promise for this, I now simply schedule a
microtask to run the code in question.

As a result, the JS global object now gets constructed at the same time
it used to (when the test calls evaluateJavaScript) and the test passes
again.

This should not change the behavior on JS side.

Test: Tools/TestWebKitAPI/Tests/WebKitCocoa/UnifiedPDFTests.mm
* Source/WebCore/page/Navigation.cpp:
(WebCore::waitForAllPromises):
(WebCore::Navigation::innerDispatchNavigateEvent):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/UnifiedPDFTests.mm:
(TestWebKitAPI::UNIFIED_PDF_TEST):

Canonical link: <a href="https://commits.webkit.org/300537@main">https://commits.webkit.org/300537@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/656f95b4e87ecc3fa2ecbdb4a0b5bdeeb02a236f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/122855 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/42567 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/33257 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/129481 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/74960 "Built successfully") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/124731 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/43286 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/51160 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/93385 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/61995 "Passed tests") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/125807 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/34512 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/109973 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/74011 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/33489 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/28130 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/72983 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/104212 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/28343 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/132210 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/49800 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/37915 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/101892 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/50177 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/106187 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/101759 "Found 2 new API test failures: TestWTF:CompletionHandlerDeathTest.MainThreadHandlerOnThreadAssertsDeathTest, TestWTF:CompletionHandlerDeathTest.ConstructionThreadHandlerOnThreadAssertsDeathTest (failure)") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25872 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/47131 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/25314 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/46572 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/49658 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/55410 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/49125 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/52476 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/50807 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

pvollan commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -865,8 +865,13 @@ struct AwaitingPromiseData : public RefCounted<AwaitingPromiseData> {
 
 // https://webidl.spec.whatwg.org/#wait-for-all
-static void waitForAllPromises(const Vector<RefPtr<DOMPromise>>& promises, Function<void()>&& fulfilledCallback, Function<void(JSC::JSValue)>&& rejectionCallback)
+static void waitForAllPromises(Document& document, const Vector<RefPtr<DOMPromise>>& promises, Function<void()>&& fulfilledCallback, Function<void(JSC::JSValue)>&& rejectionCallback)
```
Can `document` be `const`?

cdumez replied:
```
CheckedRef<EventLoopTaskGroup> checkedEventLoop();
```
Not currently marked as `const` at least.

cdumez committed:
REGRESSION (299927@main): [ macOS ]TestWebKitAPI.UnifiedPDF.PasswordFormShouldDismissAfterNavigation is a consistent failure
https://bugs.webkit.org/show_bug.cgi?id=298987
rdar://160728656

Reviewed by Per Arne Vollan and Basuke Suzuki.

The code in Navigation::innerDispatchNavigateEvent() was getting called from
FrameLoader::commitProvisionalLoad() and would end up constructing the JS
global object too early and the test's injected bundle would fail to setup
`window.internals`. The reason innerDispatchNavigateEvent() would end up
constructing the JS global object was due to constructing a dummy resolved
JS promise, for the sole purpose of scheduling work in a microtask.
Instead of constructing a JS promise for this, I now simply schedule a
microtask to run the code in question.

As a result, the JS global object now gets constructed at the same time
it used to (when the test calls evaluateJavaScript) and the test passes
again.

This should not change the behavior on JS side.

Test: Tools/TestWebKitAPI/Tests/WebKitCocoa/UnifiedPDFTests.mm
* Source/WebCore/page/Navigation.cpp:
(WebCore::waitForAllPromises):
(WebCore::Navigation::innerDispatchNavigateEvent):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/UnifiedPDFTests.mm:
(TestWebKitAPI::UNIFIED_PDF_TEST):

Canonical link: https://commits.webkit.org/300537@main
